src/date-range-picker/use-date-range-picker.tsx

```diff
@@ -0,0 +1,106 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+import { useEffect, useState, useRef } from 'react';
+import { DateRangePickerProps, Focusable } from './interfaces';
+import { setTimeOffset } from './time-offset';
+
+const VALID_RANGE: DateRangePickerProps.ValidRangeResult = { valid: true };
+
+/**
+ * This function fills in a start and end time if they are missing.
+ */
+function fillMissingTime(value: DateRangePickerProps.AbsoluteValue | null) {
+  if (!value) {
+    return value;
+  }
+  const [startDate, startTime] = value.startDate.split('T');
+  const [endDate, endTime] = value.endDate.split('T');
+  return {
+    ...value,
+    startDate: startTime ? value.startDate : `${startDate}T00:00:00`,
+    endDate: endTime ? value.endDate : `${endDate}T23:59:59`,
+  };
+}
+
+function formatValue(
+  value: null | DateRangePickerProps.Value,
+  { timeOffset, dateOnly }: { timeOffset: number; dateOnly: boolean }
+): null | DateRangePickerProps.Value {
+  if (!value || value.type === 'relative') {
+    return value;
+  }
+  if (dateOnly) {
+    return {
+      type: 'absolute',
+      startDate: value.startDate.split('T')[0],
+      endDate: value.endDate.split('T')[0],
+    };
+  }
+  return setTimeOffset(value, timeOffset);
+}
+
+function getDefaultMode(
+  value: null | DateRangePickerProps.Value,
+  relativeOptions: readonly DateRangePickerProps.RelativeOption[],
+  rangeSelectorMode: DateRangePickerProps.RangeSelectorMode
+) {
+  if (value && value.type) {
+    return value.type;
+  }
+  if (rangeSelectorMode === 'relative-only') {
+    return 'relative';
+  }
+  if (rangeSelectorMode === 'absolute-only') {
+    return 'absolute';
+  }
+  return relativeOptions.length > 0 ? 'relative' : 'absolute';
+}
+
+export function useDateRangePicker({ value, relativeOptions, rangeSelectorMode }: UseDateRangePickerProps) {
+  const [rangeSelectionMode, setRangeSelectionMode] = useState<'absolute' | 'relative'>(
+    getDefaultMode(value, relativeOptions, rangeSelectorMode)
+  );
+
+  useEffect(() => {
+    setRangeSelectionMode(getDefaultMode(value, relativeOptions, rangeSelectorMode));
```

Comment:
Another anti-pattern from the same article: https://beta.reactjs.org/learn/you-might-not-need-an-effect#updating-state-based-on-props-or-state